### ISSUE TITLE
Rename Class as FileName

### ISSUE THREAD

**Issue by Giggiux** [scalameta/metals-feature-requests#201]
**Is your feature request related to a problem? Please describe.**

When there is a single Class in a file that doesn't have the same name as the class, It would be useful to have a code action to rename the Class as the FileName, being that is good practice doing so.

**Describe the solution you'd like**

Add a code action that renames the Class and all its references.

**Describe alternatives you've considered**

As suggested in scalameta/metals#2691 another alternative to the suggested solution (implemented in scalameta/metals#2723) is to do the opposite: rename the file as the class. This doesn't always satisfy the need. 

**Additional context**

How to implement this was discussed with @tgodzik, and we came up with a possible solution using the `RenameProvider`.
Being a sort of expensive operation, the idea is to start supporting [`codeAction/resolve`](https://microsoft.github.io/language-server-protocol/specification#codeAction_resolve) requests, by extending the current `CodeAction` class with a `resolve` method. This default method will just return the same `CodeAction` received in the request. Method to be override in the Code Action implementations, if required.
There is to take in consideration that not all clients could have the "`codeAction.resolveSupport` `edit` property" announced, so an idea would be to implement a fallback for those operations as a ServerCommand.

 

**Search terms**:

rename class codeAction/resolve

**Comment by kubukoz** [comment]
I like the idea, but does it have to be a full rename as well? Just a linting rule (scalafix syntactic rule?) should be good enough to point out that a file doesn't contain a class with the same name, then you can rename it using the existing renaming functionality.

**Comment by tgodzik** [comment]
> I like the idea, but does it have to be a full rename as well? Just a linting rule (scalafix syntactic rule?) should be good enough to point out that a file doesn't contain a class with the same name, then you can rename it using the existing renaming functionality.

We don't yet get any diagnostics from Scalafix, that one still needs to be worked on. And also user would need to add that rule themselves, which doesn't seem to contain that much value. Maybe we could think of some kind of `MetalsRules` that would be added as a default, which would do a bunch of linting? 

It would be nicer to see a warning in the source file and just fix it there. But I still believe having a code action for that is a good thing to have.

**Comment by kubukoz** [comment]
> Maybe we could think of some kind of MetalsRules that would be added as a default, which would do a bunch of linting?

I like the idea. But in general I would argue that rules like this don't _need_ to be an IDE feature - it could be completely delegated to the linter, with the only work for Metals being the actual integration (displaying diagnostics from Scalafix in the future).

> It would be nicer to see a warning in the source file and just fix it there. But I still believe having a code action for that is a good thing to have.

Sure, just trying to help you avoid scope creep 😅

**Comment by tgodzik** [comment]
> > Maybe we could think of some kind of MetalsRules that would be added as a default, which would do a bunch of linting?
> 
> I like the idea. But in general I would argue that rules like this don't _need_ to be an IDE feature - it could be completely delegated to the linter, with the only work for Metals being the actual integration (displaying diagnostics from Scalafix in the future).

Makes sense, we don't plan doing any specific new linting diagnostics in Metals, we would delegate it to scalafix for sure. We might just implement some additional rule that would be included by default in `.scalafix.conf`, but this is just an idea for now. Code actions are a bit different since they can work on a range, which is not the case currently for Scalafix (works on the whole file). We could try doing the code actions in response to warnings coming from scalafix, but I don't think we can enforce everyone to use scalafix, so it might be best to have separate code actions anyway,
